python/test/test_api.py

```diff
@@ -0,0 +1,238 @@
+    def test_gpu_memory(self):
+        import cupy
+
+        server = tritonserver.Server(
+            model_repository="/workspace/models", exit_timeout=5
+        )
+
+        server.start(blocking=True)
+
+        test = server.get_model("test")
+        fp16_input = cupy.array([[5], [6], [7], [8]], dtype=numpy.float16)
+        responses = test.infer(inputs={"fp16_input": fp16_input}, request_id="1")
+
+        for response in responses:
+            print(response)
+
+        responses = server.models["test"].infer(
+            inputs={"fp16_input": fp16_input}, request_id="1"
+        )
+
+        for response in responses:
+            print(response)
+        try:
+            pass
+        #            server.stop()
+        except Exception as error:
+            print(error)
+
+    def test_unload(self):
+        server = tritonserver.Server(
+            model_repository="/workspace/models",
+            exit_timeout=5,
+            model_control_mode=tritonserver.ModelControlMode.EXPLICIT,
+            startup_models=["test"],
+            log_verbose=True,
+            log_error=True,
+            log_info=True,
+        )
+        server.start(blocking=True)
+
+        model = server.models["test"]
+
+        responses = model.infer(
+            inputs={"fp16_input": numpy.array([[0.5]], dtype=numpy.float16)}
+        )
+
+        print(list(responses)[0])
+
+        print(model.is_ready())
+
+        model_versions = [key for key in server.models.keys() if key[0] == model.name]
+
+        server.unload_model(model.name, blocking=True)
+
+        server.unload_model("foo", blocking=True)
+
+        #        del responses
+
+        while True:
+            if [
+                key
+                for key in model_versions
+                if (
+                    server.models[key].state not in server.models[key].state is not None
+                    and server.models[key].state != "UNAVAILABLE"
+                )
+            ]:
+                print(list(server.models.items()))
+                time.sleep(5)
+                continue
+            break
+        print(server.models[model.name])
+        print(list(server.models.items()))
+
+        print(model.is_ready())
+        server.stop()
+
+    def test_inference(self):
+        server = tritonserver.Server(
+            model_repository="/workspace/models",
+            exit_timeout=5
+            #           log_verbose=True,
+        )
+        #            log_error=True,
+        server.start()
+        while not server.is_ready():
+            pass
+
+        response_queue = queue.SimpleQueue()
+
+        test = server.get_model("test")
+        test_2 = server.get_model("test_2")
+
+        inputs = {
+            "text_input": numpy.array(["hello"], dtype=numpy.object_),
+            "fp16_input": numpy.array([["1"]], dtype=numpy.float16),
+        }
+
+        responses_1 = test.infer(
```

Comment:
Why does infer call returns an iterator even when  response_queue is specified? 
If the responses are directed to the response_queue, then infer should not return the iterator.
Is there a way to merge the individual iterators into a single combined iterator object instead of passing a response queue as an explicit option?

Reply:
I kept both the iterator and response queue to keep the API consistent for cancellation. Felt that was better than having the return type be different in this case between the two.

In terms of merging the iterators - it can be done but you'd have to check 'if async do this' on each response - as they need different handling of the queues (async requires to push to the event loop thread).

thought it would be cleanest to keep them seperated for now

Reply:
I'm leaning toward just removing the "queue" functionality for beta and see whether there are any requests for it. If I remember correctly, from our previous conversation it sounded like the use-case was niche (i.e., to provide an ability for processing multiple responses in parallel).